### ISSUE TITLE
using host value that includes path doesn't parse options correctly

### ISSUE THREAD

**Issue by EvaLok** [node-influx/node-influx#156]
encountered unexpected behaviour when trying to use module with kubernetes proxy address to influxdb.

initialising `Influx` object with a proxy host name doesn't work, eg:

```js
var influxClient = Influx( {
    "hosts": [
      "http://localhost:8001/api/v1/proxy/namespaces/default/services/influxdb:8086"
    ]
    database: 'somedatabase'
});
```

.. results in incorrect database parameter being set:
```js
{
  "host": "localhost",
  "port": "8001",
  "protocol": "http:",
  "database": "api/v1/proxy/namespaces/default/services/influxdb:8086"
}
```

.. should be:
```js
{
  "host": "localhost:8001/api/v1/proxy/namespaces/default/services/influxdb:8086",
  "port": "8001",
  "protocol": "http:",
  "database": "somedatabase"
}
```

.. problem seems to be in the parsing here:
https://github.com/node-influx/node-influx/blob/master/index.js#L19-L38

**Comment by bencevans** [comment]
I'm not familiar with the inner workings of Kubernetes. This client works given a direct address to the InfluxData server.

Is there a way you can get the IP/hostname and port for communicating with the InfluxData server directly or must you use this proxy with uncommon url path? If you can, gain this information and provide that to this module instead.

Otherwise you will need to change the parsing as mentioned before within the client. A pull-request is welcome.

**Comment by EvaLok** [comment]
yes, it is working with direction connections, but the proxying is very useful for development purposes.. i'll work on a PR for it

**Comment by EvaLok** [comment]
hmm.. on further inspection, it seems like the host data should also include the path in addition to the host; this will be a less than very simple fix, so unfortunately i don't have time for it this week, but hopefully next week or so i'll be able to get to it

**Comment by SPARTAN563** [comment]
Before looking at implementing a parser for those kinds of URLs (which violate the [URL specification](https://url.spec.whatwg.org/#urls), specifically that a host is defined as an IPv4, IPv6 or domain name and that it is **followed** by an optional port specifier), it may be worth investigating the use of [percent encoding](https://url.spec.whatwg.org/#percent-encode) for the portions of the "domain name" which are not truly valid domain name characters.

For example, use this instead and see whether it works:

```javascript
var influxClient = Influx( {
    "hosts": [
      "http://localhost%3A8001%2Fapi%2Fv1%2Fproxy%2Fnamespaces%2Fdefault%2Fservices%2Finfluxdb%3A8086:8001"
    ]
    database: 'somedatabase'
});
```

To summarize:
 - Replace instances of `:` with `%3A`
 - Replace instances of `/` with `%2F`
 - Append the port number to the end of the new URL

This can be achieved by using `encodeURIComponent()` in the browser and Node.js.

I'll be interested to hear whether that works, I can imagine reasons why it may not (specifically if the DNS resolver in Node.js does not handle hosts with invalid domain names correctly).

All that being said, I agree with @bencevans that this is something best handled by a resolver that sits between your application and the influxdb client library, it is platform specific and really not something the library should need to know about.

**Comment by EvaLok** [comment]
no luck with the encoding, i suppose since you guys consider this to be out of the project scope, i will look into other workarounds; thanks for the help in any case :)

**Comment by EvaLok** [comment]
@SPARTAN563 the example given might be a bit misleading, I believe that the port is actually still `8001`, the `8086` is just part of the path. i think this is probably valid, since this:

`http://localhost:8001/api/v1/proxy/namespaces/default/services/influxdb:8086/query?..query-stuff`

works in chrome (returns expected results for a query in JSON)

i'm not very experienced with this stuff, but would it still be out of scope to allow a path to be passed as well as the host?

**Comment by SPARTAN563** [comment]
I see, was under the assumption that you were looking for the full `localhost:8001/api/v1/proxy/namespaces/default/services/influxdb:8086` to be parsed as the host component (in other words, a custom URL parsing library).

Internally this library does make use of `url.parse()` however the format it works with is `<protocol>://<host>[:<port>]/[<database>]` - so one would need to modify the code to only adopt the path component as the database name if there wasn't a database specified already.

In addition to that, you would need to update the `lib/InfluxRequest.js` file to include the `path` component for each of the host entries, as it currently does not. Next step would be to add a couple of tests showing that it works as expected, and preferably some way to test them on Travis CI (which doesn't have access to a proxy with that kind of URL).

**Comment by ericvrp** [comment]
Has anyone made progress on this? I have a usecase where nginx runs a reverse proxy (on https://mydomain.com/influx) to http://localhost:8086

**Comment by NateZimmer** [comment]
I'm running into the same issue... I already rolled my own equivalent of this then I tried to use this library so I wouldn't be 're-inventing the wheel' only to find that it didn't handle basic URLs? 

I got a https://myDomain.com/someText/ forwarded to influx via nginx. This is quite common for servers where they will be running multiple web-apps all forwarded to different urls.
